### DISCUSSION TITLE
Manual multiply-adds: inference cost calculated automatically?

### DISCUSSION THREAD

**Discussion by rajb245** [Xilinx/brevitas-radioml-challenge-21#12] [Export & Inference Cost]
Lets say I manually do an operation like `(a*b).sum(dim=something)` on torch tensors `a` and `b`. This clearly incurs multiply-adds but isn't cleanly implemented as a matmul or conv or something. I also have no idea what this looks like when exported to ONNX. Does the inference cost stuff in finn know how to automatically calculate the cost of this?

**Comment by maltanar** [comment]
Hi Raj, our current inference cost measurement implementation only considers the MACs from convolutional and FC layers. We additionally declare several ops as zero-cost, which includes individual `Mul`  and `Add` ONNX nodes:

https://github.com/Xilinx/finn-base/blob/feature/itu_competition_21/src/finn/analysis/inference_cost.py#L208

The intention here is to focus on the layers that traditionally make up the largest (typically 90%+) part of the MAC/parameter budget of the inference.

**Comment by rajb245** [comment]
Now that the challenge is over: if some operations have zero cost, and lowering the cost is the primary objective of the challenge (subject to an accuracy constraint), then its logical to try to construct networks that use mostly (or only) zero-cost operations. I alluded to this above, but It is possible to make the usual primitives, like linear and conv layers, out of only these zero-cost operations with a few lines of code, and operations like `(a*b).sum(dim=...)` are at the heart of the calculation. I was able to get sensible graphs with a cost of zero according to finn, but I did not bother training such things because I understood such a thing is not in the spirit of the competition.
